JOSS paper 2023 (#127)
trixi-framework/libtrixi: paper.md

---
title: '[libtrixi]{.sc}: an interface library for using Trixi.jl from C/C++/Fortran'
tags:
  - numerical simulation
  - Julia
  - cross-language interoperability
  - computational fluid dynamics (CFD)
  - high performance computing (HPC)
authors:
  - name: Michael Schlottke-Lakemper
    orcid: 0000-0002-3195-2536
    corresponding: true
    affiliation: "1, 2" # (Multiple affiliations must be quoted)
  - name: Benedict Geihe

>   - orcid: 0000-0002-9519-6872

Reviewer: ```suggestion
    orcid: 0000-0002-9519-6872
```

    affiliation: 3
  - name: Gregor J. Gassner
    orcid: 0000-0002-1752-1158
    affiliation: 3
affiliations:
 - name: Applied and Computational Mathematics, RWTH Aachen University, Germany
   index: 1
 - name: High-Performance Computing Center Stuttgart (HLRS), University of Stuttgart, Germany
   index: 2
 - name: Department of Mathematics and Computer Science, University of Cologne, Germany
   index: 3
date: 01 October 2023
bibliography: paper.bib
---

# Summary

> [`Libtrixi`](https://github.com/trixi-framework/libtrixi) [@schlottkelakemper2023libtrixi]

Reviewer: Capitalized `Libtrixi` here because it is the beginning of a sentence, right?
Non-capitalized in title (line 2) because that is the way it is really written?
(I'm fine with that!)

the C/Fortran interface of `libtrixi`.

With `libtrixi` it is possible to use the modern, high-order discretization schemes
implemented in Trixi.jl within existing frameworks or legacy applications. Since Julia has
similar high-performance properties as C/C++/Fortran, this approach facilitates the use of
the rapid prototyping capabilities of Julia for developing new mathematical models or
numerical methods, while continuing to use established codes for the overall process
control. Both `libtrixi` and Trixi.jl are available under the MIT license.

# Statement of need

Numerical simulations of conservation laws are used to accurately predict many naturally
occurring processes in various areas of physics, such as fluid flow, astrophysics, earth
systems, or weather and climate modeling. These phenomena characteristically exhibit a broad
range of spatial and temporal length scales, making it necessary to use finely resolved
computational grids. Therefore, high-performance computing (HPC) techniques are required
to render the numerical solution feasible on large-scale compute systems.

Consequently, many simulation tools are written in traditional HPC languages such as C, C++,

> or Fortran, e,g., deal.II [@dealii2019design] or PETSc [@petsc-efficient]. These languages

Reviewer: ```suggestion
or Fortran, e.g., deal.II [@dealii2019design] or PETSc [@petsc-efficient]. These languages
```

offer high computational performance, but often at the cost of being complex to learn and
maintain. On the other hand, languages like Python, which are more amenable for rapid
prototyping or less experienced users, are usually not fast enought to use without
specialized packages that utilize kernels written in a compiled language. For example,
Python's well-known NumPy library [@2020NumPy-Array] has its performance-critical code
implemented in C.

The Julia programming language [@Bezanson_Julia_A_fresh_2017] aims to provide a new approach

> to scientific computing. It strives to solve the two-language problem by providing an

Reviewer: Is it clear that "two-language problem" is a fixed term which exactly refers to the problem just described in the preceding paragraph?

Author: You're right. It is not clear enough, and I hope it is clearer with 724588c.

This control direction, where a Julia library is managed from C/Fortran, has been rarely
used in practice so far. Besides making Trixi.jl available to a wider scientific audience,
`libtrixi` is thus a research project to investigate the efficacy of using Julia-based
libraries in existing code environments, and can eventually serve as a blueprint for
other similar efforts. Questions such as how to retain the flexibility of Julia while
providing a traditional, fixed API, how to use system-local third-party libraries, or how to
interact with Fortran are investigated and answered.

# Technical overview
`Libtrixi` consists of three main parts:

- the Julia package
  [LibTrixi.jl](https://github.com/trixi-framework/libtrixi/tree/main/LibTrixi.jl), which
  provides a traditional library API to Trixi.jl in Julia,
- a C API that exposes this Julia API as an ordinary, shared C library,
- Fortran bindings for the C API.

\autoref{fig:libtrixi_overview} illustrates the general workflow: A main program written in
C/C++/Fortran interacts with the public-facing C or Fortran API of `libtrixi`. The
Fortran API is little more than a set of language bindings, with some extra code to handle

> the conversion between Fortran strings and C strings. The C API is slightly more involved,

Reviewer: It also handles "logical functions" (?)
So write more generally "data types" or just use strings as an example
(I'm fine with with every variant)

Author: Good catch! Fixed in 2488874.

providing functionality to initialize and eventually finalize the Julia runtime. During
initialization, C function pointers are obtained from Julia to the relevant API functions
implemented in LibTrixi.jl. Most of the C API then just forwards the API calls to these
function pointers.

Finally, LibTrixi.jl is the actual library layer. Since Trixi.jl uses a composable design
and relies heavily on Julia's type system, it is necessary to repackage this flexibility
such that it can be exposed in a traditional API with static types. This is achieved by
converting Trixi.jl's *elixirs*, which consist of all code to set up and run a simulation,
into so-called *libelixirs*. These libelixirs are used to intialize a simulation state,
which is then stored internally in LibTrixi.jl and assigned a unique integer. This integer
handle is exposed in the C/Fortran API and facilitates all interaction between the main

> program and the actual simulation.

Reviewer: It also allows to control multiple simulations concurrently (which however is not tested and therefore should probably not be mentioned?)

> ![Overview of a main program interacting with Trixi.jl through

Reviewer: Nice sketch!
```suggestion
![Overview of main programs interacting with Trixi.jl through
```

Author: I've modified the figure and caption slightly in 23f60dd.

`libtrixi`.\label{fig:libtrixi_overview}](libtrixi-overview.pdf)

As an alternative to the aforementioned translation layer written in C, which exposes the
Julia API of LibTrixi.jl via function pointers, there exists experimental support in
`libtrixi` for compiling LibTrixi.jl directly into a C library. This is achieved with
PackageCompiler.jl[^1], a Julia package that allows one to compile Julia packages directly
into a shared C library or even a standalone executable. While other Julia packages exist

> that provide build options for PackageCompiler.jl, they typically use it offer

Reviewer: ```suggestion
that provide build options for PackageCompiler.jl, they typically use it to offer
```

Author: 21519ac
